File: tennis-game/kotlin-arrow-tennis-game/src/main/kotlin/Console.kt
import arrow.fx.IO

object Console {
    val ask: (String) -> IO<String> = { question -> puts(question).flatMap { reads } }

    val puts: (String) -> IO<Unit> = { str -> IO { println(str) } }

    private val reads: IO<String> = IO { readLine()!! }
}

File: tennis-game/swift-bow-tennis-game/swift-bow-tennis-game/Console.swift
//
//  Console.swift
//  swift-bow-tennis-game
//
//  Created by Fabrizio Duroni on 08.09.20.
//  Copyright © 2020 Fabrizio Duroni. All rights reserved.
//

import Foundation
import BowEffects

func puts(message: String) -> IO<Never, Void> {
    return IO.invoke { print(message) }
}

func reads() -> IO<Never, String> {
    return IO.invoke { readLine()! }
}

func ask(question: String) -> IO<Never, String> {
    return puts(message: question).flatMap(reads)^
}
